File: src/Database/Python/Tests/vector_property_test.py
# -*- coding: utf-8 -*-
__copyright__ = """This code is licensed under the 3-clause BSD license.
Copyright ETH Zurich, Laboratory of Physical Chemistry, Reiher Group.
See LICENSE.txt for details.
"""

import scine_utilities as utils
import scine_database as db
import unittest
import sys
import os
from typing import Dict, Any


class VectorPropertyTest(unittest.TestCase):
    def setUp(self):
        self.manager = db.Manager()
        self.manager.credentials.hostname = os.environ.get(
            'TEST_MONGO_DB_IP', '127.0.0.1')
        self.manager.credentials.port = int(os.environ.get('TEST_MONGO_DB_PORT', 27017))
        self.manager.credentials.database_name = "unittest_db_VectorPropertyTest"
        self.manager.connect()
        self.manager.init()

    def tearDown(self):
        self.manager.wipe()

    def test_make_one(self):
        # Setup
        coll = self.manager.get_collection("properties")
        c = db.ID()
        s = db.ID()
        model = db.Model("dft", "pbe", "def2-svp")
        test = db.VectorProperty.make("density_matrix", model, [1.234, 42.321], s, c, coll)
        assert test.has_id()

        # Check Fields
        model = test.get_model()
        comment = test.get_comment()
        name = test.get_property_name()
        calculation = test.get_calculation()
        structure = test.get_structure()
        data_db = test.get_data()

        assert model.method == "pbe"
        assert model.basis_set == "def2-svp"
        assert model.spin_mode == "any"
        assert comment == ""
        assert name == "density_matrix"
        assert calculation.string() == c.string()
        assert structure.string() == s.string()
        assert all(data_db == [1.234, 42.321])

    def test_make_two(self):
        # Setup
        coll = self.manager.get_collection("properties")
        model = db.Model("dft", "pbe", "def2-svp")
        test = db.VectorProperty.make("density_matrix", model, [1.234, 42.321], coll)
        assert test.has_id()

        # Check Fields
        model = test.get_model()
        comment = test.get_comment()
        name = test.get_property_name()
        data_db = test.get_data()

        assert model.method == "pbe"
        assert model.basis_set == "def2-svp"
        assert model.spin_mode == "any"
        assert comment == ""
        assert name == "density_matrix"
        assert not test.has_calculation()
        assert not test.has_structure()
        assert all(data_db == [1.234, 42.321])

    def test_create_one(self):
        # Setup
        coll = self.manager.get_collection("properties")
        c = db.ID()
        s = db.ID()
        model = db.Model("dft", "pbe", "def2-svp")
        test = db.VectorProperty()
        test.link(coll)
        test.create(model, "density_matrix", s, c, [1.234, 42.321])
        assert test.has_id()

        # Check Fields
        model = test.get_model()
        comment = test.get_comment()
        name = test.get_property_name()
        calculation = test.get_calculation()
        structure = test.get_structure()
        data_db = test.get_data()

        assert model.method == "pbe"
        assert model.basis_set == "def2-svp"
        assert model.spin_mode == "any"
        assert comment == ""
        assert name == "density_matrix"
        assert calculation.string() == c.string()
        assert structure.string() == s.string()
        assert all(data_db == [1.234, 42.321])

    def test_create_two(self):
        # Setup
        coll = self.manager.get_collection("properties")
        model = db.Model("dft", "pbe", "def2-svp")
        test = db.VectorProperty()
        test.link(coll)
        test.create(model, "density_matrix", [1.234, 42.321])
        assert test.has_id()

        # Check Fields
        model = test.get_model()
        comment = test.get_comment()
        name = test.get_property_name()
        data_db = test.get_data()

        assert model.method == "pbe"
        assert model.basis_set == "def2-svp"
        assert model.spin_mode == "any"
        assert comment == ""
        assert name == "density_matrix"
        assert not test.has_calculation()
        assert not test.has_structure()
        assert all(data_db == [1.234, 42.321])

    def test_data(self):
        # Setup
        coll = self.manager.get_collection("properties")
        s = db.ID()
        c = db.ID()
        model = db.Model("dft", "pbe", "def2-svp")
        test = db.VectorProperty.make("density_matrix", model, [1.234, 42.321], s, c, coll)
        assert test.has_id()

        assert all(test.get_data() == [1.234, 42.321])
        test.set_data([0.123, 4.567, 1000.00])
        assert all(test.get_data() == [0.123, 4.567, 1000.00])

    def test_data_failure(self):
        # Setup
        coll = self.manager.get_collection("properties")
        test = db.VectorProperty()
        self.assertRaises(RuntimeError, lambda: test.set_data([1.234, 42.321]))
        self.assertRaises(RuntimeError, lambda: test.get_data())
        test.link(coll)
        self.assertRaises(RuntimeError, lambda: test.set_data([1.234, 42.321]))
        self.assertRaises(RuntimeError, lambda: test.get_data())

File: src/Database/Tests/VectorPropertyTest.cpp
/**
 * @file
 * @copyright This code is licensed under the 3-clause BSD license.\n
 *            Copyright ETH Zurich, Department of Chemistry and Applied Biosciences, Reiher Group.\n
 *            See LICENSE.txt for details.
 */
#include <Database/Exceptions.h>
#include <Database/Manager.h>
#include <Database/Objects/Model.h>
#include <Database/Objects/VectorProperty.h>
#include <gmock/gmock.h>

using namespace testing;
namespace Scine {
namespace Database {
namespace Tests {
/**
 * @class VectorPropertyTest VectorPropertyTest.cpp
 * @brief Comprises tests for the class Scine::Database::VectorProperty.
 * @test
 */
class VectorPropertyTest : public Test {
 public:
  VectorPropertyTest() {
  }
  static Credentials credentials;
  static Manager db;

  static void SetUpTestCase() {
    db.setCredentials(credentials);
    db.connect();
    db.init();
  }
  static void TearDownTestCase() {
    db.wipe();
  }
};
Credentials VectorPropertyTest::credentials(TEST_MONGO_DB_IP, std::atoi(TEST_MONGO_DB_PORT), "unittest_db_VectorPropertyTest");
Manager VectorPropertyTest::db;

TEST_F(VectorPropertyTest, Create1) {
  // Setup
  auto coll = db.getCollection("properties");
  Eigen::VectorXd data(4);
  data << 42.0, 24.0, 40.2, 2.04;
  ID s, c;
  VectorProperty test = VectorProperty::create("density_matrix", Model("dft", "pbe", "def2-svp"), data, s, c, coll);
  ASSERT_TRUE(test.hasId());

  // Check Fields
  auto model = test.getModel();
  auto comment = test.getComment();
  auto name = test.getPropertyName();
  auto calculation = test.getCalculation();
  auto structure = test.getStructure();
  auto data_db = test.getData();

  ASSERT_EQ(model.method, "pbe");
  ASSERT_EQ(model.basisSet, "def2-svp");
  ASSERT_EQ(model.spinMode, "any");
  ASSERT_EQ(comment, "");
  ASSERT_EQ(name, "density_matrix");
  ASSERT_EQ(calculation, c);
  ASSERT_EQ(structure, s);
  ASSERT_EQ(data_db[0], 42.0);
  ASSERT_EQ(data_db[1], 24.0);
  ASSERT_EQ(data_db[2], 40.2);
  ASSERT_EQ(data_db[3], 2.04);
}

TEST_F(VectorPropertyTest, Create2) {
  // Setup
  auto coll = db.getCollection("properties");
  Eigen::VectorXd data(4);
  data << 42.0, 24.0, 40.2, 2.04;
  VectorProperty test = VectorProperty::create("density_matrix", Model("dft", "pbe", "def2-svp", "restricted"), data, coll);
  ASSERT_TRUE(test.hasId());

  // Check Fields
  auto model = test.getModel();
  auto comment = test.getComment();
  auto name = test.getPropertyName();
  auto data_db = test.getData();

  ASSERT_EQ(model.method, "pbe");
  ASSERT_EQ(model.basisSet, "def2-svp");
  ASSERT_EQ(model.spinMode, "restricted");
  ASSERT_EQ(comment, "");
  ASSERT_EQ(name, "density_matrix");
  ASSERT_FALSE(test.hasCalculation());
  ASSERT_FALSE(test.hasStructure());
  ASSERT_EQ(data_db[0], 42.0);
  ASSERT_EQ(data_db[1], 24.0);
  ASSERT_EQ(data_db[2], 40.2);
  ASSERT_EQ(data_db[3], 2.04);
}

TEST_F(VectorPropertyTest, Data) {
  // Setup
  auto coll = db.getCollection("properties");
  Eigen::VectorXd data = Eigen::VectorXd::Zero(4);
  VectorProperty test = VectorProperty::create("density_matrix", Model("dft", "pbe", "def2-svp"), data, coll);
  ASSERT_TRUE(test.hasId());

  auto data_db = test.getData();
  ASSERT_EQ(data_db[0], 0.0);
  ASSERT_EQ(data_db[1], 0.0);
  ASSERT_EQ(data_db[2], 0.0);
  ASSERT_EQ(data_db[3], 0.0);
  // clang-format off
  data << 42.0, 24.0, 40.2, 2.04;
  // clang-format on
  test.setData(data);
  data_db = test.getData();
  ASSERT_EQ(data_db[0], 42.0);
  ASSERT_EQ(data_db[1], 24.0);
  ASSERT_EQ(data_db[2], 40.2);
  ASSERT_EQ(data_db[3], 2.04);
}

TEST_F(VectorPropertyTest, DataFails) {
  // Setup
  auto coll = db.getCollection("properties");
  VectorProperty test;
  Eigen::VectorXd data;

  ASSERT_THROW(test.setData(data), Exceptions::MissingLinkedCollectionException);
  ASSERT_THROW(test.getData(), Exceptions::MissingLinkedCollectionException);
  test.link(coll);
  ASSERT_THROW(test.setData(data), Exceptions::MissingIDException);
  ASSERT_THROW(test.getData(), Exceptions::MissingIDException);
}

} /* namespace Tests */
} /* namespace Database */
} /* namespace Scine */
